File: mcp/py/src/jeffs_brain_memory_mcp/tools/ask.py
# SPDX-License-Identifier: Apache-2.0
"""``memory_ask`` tool definition."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from ..client import AskArgs, MemoryClient, ProgressEmitter
from ._types import ToolDef


class AskInput(BaseModel):
    query: str = Field(min_length=1, max_length=8192)
    brain: str | None = None
    top_k: int | None = Field(default=None, ge=1, le=50)


async def _handle(
    args: BaseModel,
    client: MemoryClient,
    progress: ProgressEmitter | None,
) -> dict[str, Any]:
    # TODO(next-pass): switch to streaming once the Python SDK surfaces a
    # token-level `ask` iterator. Today the local implementation returns a
    # final payload and emits two coarse progress events (retrieved +
    # answered), mirroring the TypeScript wrapper.
    assert isinstance(args, AskInput)
    return await client.ask(
        AskArgs(query=args.query, brain=args.brain, top_k=args.top_k),
        progress,
    )


ask_tool = ToolDef(
    name="memory_ask",
    description=(
        "Ask a question grounded in the brain. Streams answer tokens as MCP "
        "progress notifications and returns the final answer with citations."
    ),
    input_model=AskInput,
    handler=_handle,
)

File: mcp/ts/src/tools/ask.ts
// SPDX-License-Identifier: Apache-2.0

import { z } from 'zod'
import { type Tool, jsonContent } from './types.js'

const schema = z.object({
  query: z.string().min(1).max(8192),
  brain: z.string().optional(),
  top_k: z.number().int().min(1).max(50).optional(),
})

export const askTool: Tool<typeof schema> = {
  name: 'memory_ask',
  description:
    'Ask a question grounded in the brain. Streams answer tokens as MCP progress notifications and returns the final answer with citations.',
  inputSchema: schema,
  async handler(args, client, ctx) {
    // TODO(next-pass): once the memory SDK surfaces a streaming `ask`
    // that yields `answer_delta` frames, forward each delta through
    // `ctx.progress(counter, delta)` so MCP clients can render tokens
    // as they arrive. Today we emit two coarse progress events
    // (retrieved + answered) which is enough to satisfy the protocol.
    const result = await client.ask(
      {
        query: args.query,
        brain: args.brain,
        topK: args.top_k,
      },
      ctx?.progress,
    )
    return jsonContent(result)
  },
}
